src/sensor_interface.c

```diff
@@ -0,0 +1,241 @@
+ * 
+ * @param None
+ * @return Pointer to the device, or NULL if the device is not found.
+ */
+const struct device *get_bme280_device(void) {
+	const struct device *const dev = DEVICE_DT_GET_ANY(bosch_bme280);
+
+	if (dev == NULL) {
+		/* No such node, or the node does not have status "okay". */
+		LOG_WRN("No BME280 sensor found. Will be using simulator instead.");
+		return NULL;
+	}
+
+	if (!device_is_ready(dev)) {
+		LOG_ERR("Error: Device \"%s\" is not ready. Check the driver initialization logs for errors.",
+		       dev->name);
+		return NULL;
+	}
+
+	LOG_INF("Found BME280 device \"%s\"", dev->name);
+	return dev;
+}
+
+/* Read the I2C BME280 sensor, or simulate the sensor data if the sensor is not
+ * available. 
+ *
+ * @param dev Pointer to the device (struct device).
+ * @return None
+ */
+void update_sensor_data(const struct device *dev) {
+	// Variables to store sensor data.
+	struct sensor_value temp_value, press_value;
+	// struct sensor_value humidity_value;
+
+	// Check if the device is ready.
+	if (dev != NULL) {
+		// LOG
+		LOG_DBG("Fetching sensor data.");
+		// Read the sensor data.
+		sensor_sample_fetch(dev);
+		sensor_channel_get(dev, SENSOR_CHAN_AMBIENT_TEMP, &temp_value);
+		sensor_channel_get(dev, SENSOR_CHAN_PRESS, &press_value);
+		// sensor_channel_get(dev, SENSOR_CHAN_HUMIDITY, &humidity_value);
+
+		// Print the sensor data.
+		LOG_DBG("Temperature: %.2f C, Pressure: %.2f Pa", 
+			sensor_value_to_double(&temp_value),
+			sensor_value_to_double(&press_value));
+
+	} else {
+		// If the device is not ready, or not found, sample the sensor 
+		// with dummy simulated values.
+		// printk("INFO: Simulating sensor data.");
+
+		// Sample the sensor with dummy values.
+		temp_value.val1 = sim_temperature;
+		temp_value.val2 = sim_temperature * 2 * 1000;
+		press_value.val1 = sim_pressure;
+		press_value.val2 = sim_pressure * 2 * 1000;
+		// humidity_value.val1 = sim_humidity;
+		// humidity_value.val2 = sim_humidity * 2 * 1000;
+
+		// Decrease the dummy values.
+		sim_temperature -= 1;
+		sim_pressure -= 3;
+		// sim_humidity -= 2;
+
+		// If the dummy values are below 0, reset them.
+		if (sim_temperature < 0)
+			sim_temperature = DEFAULT_TEMPERATURE;
+		if (sim_pressure < 0)
+			sim_pressure = DEFAULT_PRESSURE;
+		// if (sim_humidity < 0)
+		//	sim_humidity = DEFAULT_HUMIDITY;
+	}
+
+	// Debug LOG
+	LOG_DBG("Temperature: %.2f C, Pressure: %.2f Pa", 
+		sensor_value_to_double(&temp_value),
+		sensor_value_to_double(&press_value)
+	//	sensor_value_to_double(&humidity_value)
+	);
+
+	// Create a struct to store the sensor readings.
+	struct sensor_value_store_t tx_data = {
+		.temperature = temp_value,
+		.pressure = press_value
+	//	.humidity = humidity_value,
+	};
+
+	size_t datatype_size = sizeof(struct sensor_value_store_t);
+	struct sensor_value_store_t *mem_ptr = k_malloc(datatype_size);
+	__ASSERT_NO_MSG(mem_ptr != 0);
+	memcpy(mem_ptr, &tx_data, datatype_size);
+
+	k_lifo_put(&sensor_value_lifo, mem_ptr);
```

Comment:
Have you tried to run it without using malloc? There has been lots of time passed since I wrote the code, but I'm remembering that it was necessary for lifo. The first implementation was as you mentioned, didn't worked.

Reply:
I didn't run the code so it can be required/limitation of the lifo implementation.